File: src/type/v2/setting.ts
/**
 * Setting型定義（v2）
 *
 * ハイブリッド詳細管理（インライン + ファイル参照）に対応した拡張型定義
 */

/**
 * 設定の種類
 */
export type SettingType = "location" | "world" | "culture" | "organization";

/**
 * 設定の詳細情報（オプショナル）
 * 各フィールドは短文インラインまたはファイル参照を選択可能
 */
export type SettingDetails = {
  /** 設定の説明（概要より詳細な説明） */
  description?: string | { file: string };
  /** 地理情報 */
  geography?: string | { file: string };
  /** 歴史 */
  history?: string | { file: string };
  /** 文化 */
  culture?: string | { file: string };
  /** 政治 */
  politics?: string | { file: string };
  /** 経済 */
  economy?: string | { file: string };
  /** 住民 */
  inhabitants?: string | { file: string };
  /** ランドマーク */
  landmarks?: string | { file: string };
};

/**
 * LSP用の検出ヒント
 */
export type SettingDetectionHints = {
  /** よく使われるパターン（例: "王都の", "王都で"） */
  commonPatterns: string[];
  /** 除外すべきパターン */
  excludePatterns: string[];
  /** 検出の信頼度（0.0～1.0） */
  confidence: number;
};

/**
 * Setting型（v2）
 *
 * 必須メタデータは型安全に管理し、詳細情報は段階的に追加可能
 */
export type Setting = {
  // ========================================
  // 必須メタデータ
  // ========================================

  /** 一意なID（プログラム的な識別子） */
  id: string;

  /** 設定名（物語内での名前） */
  name: string;

  /** 設定の種類 */
  type: SettingType;

  /** 登場するチャプターのIDリスト */
  appearingChapters: string[];

  /** 短い概要（必須） */
  summary: string;

  // ========================================
  // オプショナル情報
  // ========================================

  /** 表示名のバリエーション（例: "王都", "首都"） */
  displayNames?: string[];

  /** 詳細情報（段階的に追加可能） */
  details?: SettingDetails;

  /** 関連する設定のID */
  relatedSettings?: string[];

  /** LSP用の検出ヒント */
  detectionHints?: SettingDetectionHints;
};

File: internal/domain/setting.go
package domain

// SettingType enumerates the kind of a Setting entity.
//
// Why: TS source defines a string literal union ("location" | "world" |
// "culture" | "organization"). A string-typed const enum preserves the wire
// value when (de)serializing JSON and keeps switch statements exhaustive at
// review time without pulling in a third-party enum library.
type SettingType string

const (
	SettingTypeLocation     SettingType = "location"
	SettingTypeWorld        SettingType = "world"
	SettingTypeCulture      SettingType = "culture"
	SettingTypeOrganization SettingType = "organization"
)

// SettingDetails captures optional long-form descriptions for a Setting.
//
// Why: Each TS field is `string | { file: string }`. We model that union as
// an inline anonymous struct pointer so callers can express three states:
//   - nil          : field absent
//   - {Value: "x"} : inline literal text
//   - {File: "p"}  : reference to an external markdown file
//
// Wave-A2 will likely lift this into a shared StringOrFileRef helper across
// Character / Setting / Foreshadowing once the union appears in more places.
type SettingDetails struct {
	Description *struct {
		Value string
		File  string
	}
	Geography *struct {
		Value string
		File  string
	}
	History *struct {
		Value string
		File  string
	}
	Culture *struct {
		Value string
		File  string
	}
	Politics *struct {
		Value string
		File  string
	}
	Economy *struct {
		Value string
		File  string
	}
	Inhabitants *struct {
		Value string
		File  string
	}
	Landmarks *struct {
		Value string
		File  string
	}
}

// SettingDetectionHints provides LSP detection tuning data.
type SettingDetectionHints struct {
	CommonPatterns  []string
	ExcludePatterns []string
	Confidence      float64
}

// Setting is the Go representation of `src/type/v2/setting.ts:Setting`.
//
// Required metadata is held by value; optional fields use either pointer
// (scalar / nested struct) or nil-slice convention.
type Setting struct {
	// Required metadata.
	ID                string
	Name              string
	Type              SettingType
	AppearingChapters []string
	Summary           string

	// Optional fields.
	DisplayNames    []string
	Details         *SettingDetails
	RelatedSettings []string
	DetectionHints  *SettingDetectionHints
}
